validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/Eth1DataManager.java

```diff
@@ -0,0 +1,168 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.validator.coordinator;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+import com.google.common.primitives.UnsignedLong;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentSkipListMap;
+import tech.pegasys.artemis.datastructures.blocks.Eth1Data;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.pow.event.Eth1BlockEvent;
+import tech.pegasys.artemis.storage.events.SlotEvent;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class Eth1DataManager {
+
+  private final EventBus eventBus;
+  private final UnsignedLong genesisTime;
+
+  private NavigableMap<UnsignedLong, Eth1Data> eth1ChainCache = new ConcurrentSkipListMap<>();
+  private final UnsignedLong RANGE_CONSTANT =
+      Constants.SECONDS_PER_ETH1_BLOCK.times(Constants.ETH1_FOLLOW_DISTANCE);
+  private volatile UnsignedLong currentVotingPeriodStartTime;
+
+  static Eth1Data getEth1Data(Eth1BlockEvent eth1BlockEvent) {
+    return new Eth1Data(
+        eth1BlockEvent.getDepositRoot(),
+        eth1BlockEvent.getDepositCount(),
+        eth1BlockEvent.getBlockHash());
+  }
+
+  public Eth1DataManager(BeaconState genesisState, EventBus eventBus) {
+    this.eventBus = eventBus;
+    this.genesisTime = genesisState.getGenesis_time();
+    this.currentVotingPeriodStartTime = voting_period_start_time(genesisState.getSlot());
+    this.eventBus.register(this);
+  }
+
+  @Subscribe
+  public void onCacheEth1BlockEvent(Eth1BlockEvent eth1BlockEvent) {
+    eth1ChainCache.put(eth1BlockEvent.getBlockTimestamp(), getEth1Data(eth1BlockEvent));
+  }
+
+  @Subscribe
+  public void onSlot(SlotEvent slotEvent) {
+    UnsignedLong slot = slotEvent.getSlot();
+    UnsignedLong voting_period_start_time = voting_period_start_time(slot);
+
+    if (voting_period_start_time.equals(currentVotingPeriodStartTime)) {
+      return;
+    }
+
+    currentVotingPeriodStartTime = voting_period_start_time;
+    prune(voting_period_start_time);
+  }
+
+  public Eth1Data get_eth1_vote(BeaconState state) {
+    NavigableMap<UnsignedLong, Eth1Data> votesToConsider = getVotesToConsider();
+    Map<Eth1Data, Eth1Vote> validVotes = new HashMap<>();
+
+    int i = 0;
+    for (Eth1Data eth1Data : state.getEth1_data_votes()) {
+      if (!votesToConsider.containsValue(eth1Data)) {
+        continue;
+      }
+
+      int finalI = i;
+      Eth1Vote vote =
+          validVotes.computeIfAbsent(
+              eth1Data,
+              key -> {
+                Eth1Vote newVote = new Eth1Vote();
```

Comment:
(nit) You could pass the index into the constructor and make the index variable `final`.
```suggestion
                Eth1Vote newVote = new Eth1Vote(currentIndex);
```

Reply:
done.